File: Data Structures & Algorithms/palindromic-substrings/submission-3.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        result = len(s)

        for c in range(len(s)):
            l, r = c - 1, c + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l, r = l - 1, r + 1
                result += 1

        for c in range(len(s) - 1):
            l, r = c, c + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l, r = l - 1, r + 1
                result += 1

        return result

File: Data Structures & Algorithms/course-schedule-ii/submission-3.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        vector<int> indegree(numCourses, 0);
        for (auto& prereq : prerequisites) {
            int a = prereq[0];
            int b = prereq[1];

            adj[b].push_back(a);
            indegree[a]++;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; ++i) {
            if (indegree[i] == 0) q.push(i);
        }

        vector<int> result;
        while (!q.empty()) {
            int course = q.front();
            q.pop();
            result.push_back(course);
            for (auto neighbour : adj[course]) {
                indegree[neighbour]--;
                if (indegree[neighbour] == 0) {
                    q.push(neighbour);
                }
            }
        }

        if (result.size() == numCourses) {
            return result;
        }
        return vector<int>();
    }
};
